src/app/components/common/node-editor/editor.ts

```diff
@@ -570,3 +574,53 @@ export function applyContentReordering(nodeId: string) {
   // Bring selected node in front of other nodes for better visual clarity
   content.reorder(view.element, null);
 }
+
+export function undoAction() {
+  let secondLastActionName = getSecondLastActionName();
+  let lastActionName = getLastActionName();
+  history.undo().then(() => {
+    if (secondLastActionName == 'AddNodeAction') {
+      dock.add(newDockFilter);
+    }
+    if(lastActionName == 'RemoveNodeAction'){
+      dock.remove(newDockFilter);
+    }
+  })
+}
+
+export function redoAction() {
+  let beforeRedoSecondLastActionName = getSecondLastActionName();
+  let beforeRedoLastActionName = getLastActionName();
+  history.redo().then(() => {
+    let afterRedoSecondLastActionName = getSecondLastActionName();
+    let afterRedoLastActionName = getLastActionName();
+    if (afterRedoSecondLastActionName == 'AddNodeAction' && beforeRedoSecondLastActionName != 'AddNodeAction') {
+      dock.remove(newDockFilter);
+    }
+    if(afterRedoLastActionName == 'RemoveNodeAction' && beforeRedoLastActionName != 'RemoveNodeAction'){
+      dock.add(newDockFilter);
+    }
+  })
+}
+
+function getSecondLastActionName() {
+  let historySnapshot = history.getHistorySnapshot();
+  let historyLength = historySnapshot.length;
+  let actionName;
+  if (historyLength >= 2) {
```

Comment:
Why history length > 2 required?

Reply:
Adding a node from dock menu creates 2 entries in history array. First is for `AddNodeAction` and second is for `DragNodeAction`. so need to check the length of array before retrieving actionName.